### PR TITLE
Avoid duplicate reading for small parquet files

### PR DESCRIPTION
**Pull request by marin-ma**
For small parquet files, when loading footer, the whole file is already read to memory, no need to read file again when loading data.
Before this change:
image

After this change:
image

### CODE REVIEW

**Comment by marin-ma** [comment]
cc: @JkSelf @zhli1142015

**Comment by JkSelf** [comment]
@marin-ma Can you help to add some unit tests?

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 1]
```diff
@@ -47,12 +47,18 @@ ReaderBase::ReaderBase(
...
+  preloadFile_ = fileLength_ <= filePreloadThreshold_ ||
+      fileLength_ <= directorySizeGuess_;
   uint64_t readSize =
       preloadFile_ ? fileLength_ : std::min(fileLength_, directorySizeGuess_);
 
-  auto stream = input_->read(
-      fileLength_ - readSize, readSize, dwio::common::LogType::FOOTER);
+  std::unique_ptr<dwio::common::SeekableInputStream> stream = nullptr;
+  if (preloadFile_) {
+    stream = input_->readFile(fileLength_, dwio::common::LogType::FOOTER);
```
Comment (Yohahaha) [inline velox/dwio/parquet/reader/ParquetReader.cpp:57]: Could we use `input_->read(0, fileLength_)` instead of create a new method `readFile`?

File: velox/dwio/parquet/reader/ParquetReader.h [hunk 2]
```diff
@@ -109,16 +109,18 @@ class ReaderBase {
...
   RowTypePtr schema_;
   std::shared_ptr<const dwio::common::TypeWithId> schemaWithId_;
 
   const bool binaryAsString = false;
 
+  bool preloadFile_ = false;
+
   // Map from row group index to pre-created loading BufferedInput.
-  std::unordered_map<uint32_t, std::unique_ptr<dwio::common::BufferedInput>>
+  std::unordered_map<uint32_t, std::shared_ptr<dwio::common::BufferedInput>>
```
Comment (Yohahaha) [inline velox/dwio/parquet/reader/ParquetReader.h:123]: Why this change is needed?

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 1 above)
Comment (marin-ma) [reply velox/dwio/parquet/reader/ParquetReader.cpp:57]: `readFile` uses enqueue/load to populate underlying buffers, which is different from what `read` do.

File: velox/dwio/parquet/reader/ParquetReader.h (hunk 2 above)
Comment (marin-ma) [reply velox/dwio/parquet/reader/ParquetReader.h:123]: For small files,`input_` buffers the whole file. Thus we can store the same BufferedInput ptr for each rowgroup.

**Comment by marin-ma** [comment]
@Yuhta PTAL.

**Comment by marin-ma** [comment]
@Yuhta Could you please take a look? Thanks!

**Comment by marin-ma** [comment]
@mbasmanova Could you please take a look? We're waiting for this PR to get reviewed.

**Comment by mbasmanova** [comment]
@marin-ma I'm out next week.

@majetideepak @Yuhta Deepak, Jimmy, would you help review this PR?

**Comment by majetideepak** [comment]
@marin-ma I see that you exposed the ReaderBase interface for testing. Can we use File I/O metrics instead to check there are no duplicate reads? As you commented in the code, ReaderBase is internal only. I am adding a PR to move that from the header file to the CPP file as it is internal only.

**Comment by marin-ma** [comment]
> @marin-ma I see that you exposed the ReaderBase interface for testing. Can we use File I/O metrics instead to check there are no duplicate reads? As you commented in the code, ReaderBase is internal only. I am adding a PR to move that from the header file to the CPP file as it is internal only.

Updated UT and removed the newly added internal-only APIs. PTAL. Thanks!

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 1 above)
Comment (majetideepak) [reply velox/dwio/parquet/reader/ParquetReader.cpp:57]: `dwio::common::LogType::FILE` since we are reading the entire file?

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 3]
```diff
@@ -476,19 +482,30 @@ void ReaderBase::scheduleRowGroups(
...
   }
   for (auto counter = 0; counter < FLAGS_parquet_prefetch_rowgroups;
        ++counter) {
     if (nextGroup) {
-      if (inputs_.count(nextGroup) != 0) {
-        auto newInput = input_->clone();
-        reader.enqueueRowGroup(nextGroup, *newInput);
-        newInput->load(dwio::common::LogType::STRIPE);
-        inputs_[nextGroup] = std::move(newInput);
+      if (inputs_.count(nextGroup) == 0) {
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/ParquetReader.cpp:499]: Looks like you fixed a bug in the row group prefetching. We should fix this independently along with a test. Can you open an issue and undo this fix for this PR?

File: velox/dwio/parquet/reader/ParquetReader.h [hunk 4]
```diff
@@ -109,16 +109,18 @@ class ReaderBase {
   const uint64_t filePreloadThreshold_;
   // Copy of options. Must be owned by 'this'.
   const dwio::common::ReaderOptions options_;
-  std::unique_ptr<velox::dwio::common::BufferedInput> input_;
+  std::shared_ptr<velox::dwio::common::BufferedInput> input_;
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/ParquetReader.h:112]: Can this change be reverted to unique_ptr? Same below.

**Review by majetideepak — COMMENTED** [review]
@marin-ma Some more comments.
It would be cleaner if we can avoid the branching for `preloadFile_`. Let me take a look.

File: velox/dwio/parquet/reader/ParquetReader.cpp [hunk 5]
```diff
@@ -47,12 +47,18 @@ ReaderBase::ReaderBase(
 }
 
 void ReaderBase::loadFileMetaData() {
-  bool preloadFile_ = fileLength_ <= filePreloadThreshold_;
+  preloadFile_ = fileLength_ <= filePreloadThreshold_ ||
+      fileLength_ <= directorySizeGuess_;
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/ParquetReader.cpp:51]: also, why include `fileLength_ <= directorySizeGuess_`?

File: velox/dwio/common/BufferedInput.h [hunk 6]
```diff
@@ -122,6 +122,14 @@ class BufferedInput {
     return std::make_unique<BufferedInput>(input_, pool_);
   }
 
+  std::unique_ptr<SeekableInputStream> readFile(
```
Comment (majetideepak) [inline velox/dwio/common/BufferedInput.h:125]: I don't think API fits the BufferedInput class. This function is used only once anyway.
Let's inline this code in the call site.

**Review by majetideepak — COMMENTED** [review]
@marin-ma probably `enqueueRowGroup` should always take `input_` and return a pointer to BufferedInput.
Inside `enqueueRowGroup`, we should check if the rowgroup range is already buffered,
if it is, we return `input_`, if not we create a clone of `input_` enqueue, load, and return that. What do you think?

**Comment by marin-ma** [comment]
> @marin-ma probably `enqueueRowGroup` should always take `input_` and return a pointer to BufferedInput. Inside `enqueueRowGroup`, we should check if the rowgroup range is already buffered, if it is, we return `input_`, if not we create a clone of `input_` enqueue, load, and return that. What do you think?

Thanks for the review. I agree with your points. This can hide the check from low-level APIs and not breaking the original design. I'll see how to improve the implementation.

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 5 above)
Comment (marin-ma) [reply velox/dwio/parquet/reader/ParquetReader.cpp:51]: @zhli1142015 @JkSelf Can you help to answer this question? Thanks!

Comment (marin-ma) [reply velox/dwio/parquet/reader/ParquetReader.cpp]: @majetideepak Because in the next line, `readSize` will be `fileLength_` if `fileLength_ <= directorySizeGuess_`. In this case the whole file is loaded.
```
  uint64_t readSize =
      preloadFile_ ? fileLength_ : std::min(fileLength_, directorySizeGuess_);
```
While the code looks weird. Optimized this part of logic in the latest commit.

File: velox/dwio/parquet/reader/ParquetReader.cpp (hunk 3 above)
Comment (marin-ma) [reply velox/dwio/parquet/reader/ParquetReader.cpp]: Separated: https://github.com/facebookincubator/velox/pull/6121

**Comment by marin-ma** [comment]
@majetideepak Updated. PTAL. Thanks!

File: velox/dwio/parquet/reader/ParquetData.h [hunk 7]
```diff
@@ -172,6 +172,24 @@ class ParquetData : public dwio::common::FormatData {
     return true;
   }
 
+  // Returns the <offset, length> of the row group.
+  std::pair<int64_t, int64_t> getRowGroupRegion(uint32_t index) const {
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/ParquetData.h:176]: move the implementation to the cpp file.

File: velox/dwio/parquet/reader/StructColumnReader.h [hunk 8]
```diff
@@ -65,6 +69,10 @@ class StructColumnReader : public dwio::common::SelectiveStructColumnReader {
  private:
   dwio::common::SelectiveColumnReader* findBestLeaf();
 
+  void enqueueRowGroup0(uint32_t index, dwio::common::BufferedInput& input);
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/StructColumnReader.h:72]: let's use `enqueueRowGroup` and `loadRowGroup`.

**Review by majetideepak — COMMENTED** [review]
@marin-ma Looks much cleaner now. Thanks!
Some comments.

File: velox/dwio/parquet/reader/StructColumnReader.cpp [hunk 9]
```diff
@@ -92,12 +92,36 @@ void StructColumnReader::read(
...
 
-void StructColumnReader::enqueueRowGroup(
+std::shared_ptr<dwio::common::BufferedInput>
+StructColumnReader::enqueueRowGroup(
+    uint32_t index,
+    const std::shared_ptr<dwio::common::BufferedInput>& input) {
+  // Check whether the entire row group has already loaded. If true,
+  // reuse the input and data will be read from buffer.
+  // Or else create a new input.
+  auto newInput = isRowGroupBuffered(index, *input) ? input : input->clone();
```
Comment (majetideepak) [inline velox/dwio/parquet/reader/StructColumnReader.cpp:102]: Can simplify as
```
 if(isRowGroupBuffered(index, *input)) {
   return input;
 }
 auto newInput = input->clone();
 enqueueRowGroup(index, *newInput);
 newInput->load(dwio::common::LogType::STRIPE);
 return newInput;
```

Comment (marin-ma) [reply velox/dwio/parquet/reader/StructColumnReader.cpp:102]: Still need to call `enqueueRowGroup` if row group is buffered. It assgins stream_ of the column chunk to the buffered data.
https://github.com/facebookincubator/velox/blob/78c6d45c4ddcb97fcbac30d76fcafaf040b39556/velox/dwio/parquet/reader/ParquetData.cpp#L109

**Review by majetideepak — APPROVED** [review]
@marin-ma thanks!

**Comment by majetideepak** [comment]
@Yuhta can you take a look?
